### ISSUE TITLE
CRTP classmethod: "Could not bind method"

### ISSUE THREAD

**Issue by nuiva** [microsoft/pylance-release#1488]
<h3> Environment data </h3>

-   Language Server version: 2021.6.2
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): Python 3.9.1
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: strict

<h3> Code Snippet / Additional information </h3>

```python
from __future__ import annotations
import typing

T = typing.TypeVar("T")

class Base(typing.Generic[T]):
    @classmethod
    def construct(cls: typing.Type[T]) -> T:
        return cls()

class Derived(Base["Derived"]):
    @classmethod
    def construct(cls) -> Derived:
        return super().construct()

d: Derived = Derived.construct()
```

<h3> Expected behaviour </h3>

`Derived.construct` should not produce type errors, because the call `super().construct()` calls `Base.construct()` with `T=Derived`. This function takes the parameter `cls: Type[Derived]` (the type is necessarily correct) and returns `Derived` (which matches the return value of `Derived.construct`).

<h3> Actual behaviour </h3>

The line `return super().construct()` produces the type error
> Could not bind method "construct" because "Base[Derived]" is not assignable to parameter "cls"
>  "Base[Derived]" is incompatible with "Derived" Pylance(reportGeneralTypeIssues)

<h3> Logs </h3>

Python Language Server Log
```
[Info  - 12.33.01] Pylance language server 2021.6.2 (pyright dafa497d) starting
[Info  - 12.33.01] Server root directory: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist
[Error - 12.51.09] An internal error occurred while type checking file "m:\Programming\Python\pylance_test\main.py": RangeError: Maximum call stack size exceeded
    at validateCallArguments (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:7185:14)
    at callback (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:7566:44)
    at callback (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:3377:41)
    at Object.doForEachSubtype (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeUtils.ts:206:9)
    at expandSubtype (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:3370:13)
    at mapSubtypesExpandTypeVars (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:3393:13)
    at validateCallArguments (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:7210:28)
    at callback (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:7566:44)
    at callback (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:3377:41)
    at Object.p [as doForEachSubtype] (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeUtils.ts:206:9)
[Error - 12.51.09] Error performing analysis: RangeError: Maximum call stack size exceeded
    at validateCallArguments (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:7185:14)
    at callback (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:7566:44)
    at callback (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:3377:41)
    at Object.doForEachSubtype (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeUtils.ts:206:9)
    at expandSubtype (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:3370:13)
    at mapSubtypesExpandTypeVars (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:3393:13)
    at validateCallArguments (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:7210:28)
    at callback (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:7566:44)
    at callback (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeEvaluator.ts:3377:41)
    at Object.p [as doForEachSubtype] (c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\pyright-internal\src\analyzer\typeUtils.ts:206:9)
[Info  - 13.20.32] No configuration file found.
[Info  - 13.20.32] No pyproject.toml file found.
[Info  - 13.20.32] Setting pythonPath for service "pylance_test": "C:\Program Files\Python39\python.exe"
[Warn  - 13.20.32] stubPath m:\Programming\Python\pylance_test\typings is not a valid directory.
[Info  - 13.20.32] Assuming Python version 3.9
[Info  - 13.20.32] Assuming Python platform Windows
Search paths for m:\Programming\Python\pylance_test
# PATH variables omitted
[Info  - 13.20.32] Searching for source files
[Info  - 13.20.32] Found 1 source file
Background analysis message: setConfigOptions
Background analysis message: ensurePartialStubPackages
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: getSemanticTokens delta
[BG(1)] getSemanticTokens delta previousResultId:1624443156365 at m:\Programming\Python\pylance_test\main.py ...
[BG(1)]   parsing: m:\Programming\Python\pylance_test\main.py (91ms)
[BG(1)]   parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 1ms] (92ms)
[BG(1)]   binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\builtins.pyi (24ms)
[BG(1)]   binding: m:\Programming\Python\pylance_test\main.py (0ms)
[BG(1)]   parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\typing.pyi [fs read 1ms] (31ms)
[BG(1)]   binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\typing.pyi (8ms)
[BG(1)]   parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 1ms] (8ms)
[BG(1)]   binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (1ms)
[BG(1)]   parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 1ms] (3ms)
[BG(1)]   binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\typing_extensions.pyi (1ms)
[BG(1)]   parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\__future__.pyi [fs read 1ms] (1ms)
[BG(1)]   binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\__future__.pyi (1ms)
[BG(1)] getSemanticTokens delta previousResultId:1624443156365 at m:\Programming\Python\pylance_test\main.py (281ms)
Background analysis message: analyze
[BG(1)] analyzing: m:\Programming\Python\pylance_test\main.py ...
[BG(1)]   checking: m:\Programming\Python\pylance_test\main.py (3ms)
[BG(1)] analyzing: m:\Programming\Python\pylance_test\main.py (3ms)
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
[FG] parsing: m:\Programming\Python\pylance_test\main.py (16ms)
[FG] parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 1ms] (95ms)
[FG] binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\builtins.pyi (24ms)
[FG] binding: m:\Programming\Python\pylance_test\main.py (0ms)
Background analysis message: markFilesDirty
Background analysis message: analyze
[FG] parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\typing.pyi [fs read 1ms] (23ms)
[FG] binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\typing.pyi (6ms)
[FG] parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 0ms] (6ms)
[FG] binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (1ms)
[FG] parsing: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 0ms] (3ms)
[FG] binding: c:\Users\USER_NAME\.vscode\extensions\ms-python.vscode-pylance-2021.6.2\dist\typeshed-fallback\stdlib\typing_extensions.pyi (1ms)
[FG] parsing: c:\Program Files\Python39\Lib\typing.py [fs read 1ms] (67ms)
[FG] binding: c:\Program Files\Python39\Lib\typing.py (15ms)
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2021.6.3, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202163-23-june-2021
